File: src/iupitermag/field.py
import numpy as np

import iupitermag._core as _iu


class Field:
    _field: _iu.PyInternalField | _iu.PyCurrentSheetField

    def __init__(self):
        """
        Base class for all field types. Contains generic routines like
        calc_field, map_calc_field, trace_field, etc.
        """
        pass

    def calc_field(self, r, theta, phi):
        """
        Calculate the field strength at a location defined in a spherical
        planetocentric coordinate system.

        Args:
            r (float): Radius in planetary radii
            theta (float): Co-latitude in radians
            phi (float): Azimuth in radians

        Returns:
            field (np.array): Magnetic field components [Br, Btheta, Bphi] in a
                array.
        """
        return self._field.calc_field(r, theta, phi)

    def calc_field_xyz(self, x, y, z):
        """
        Calculate the field at a location defined by (x, y, z) in a Cartesian
        coordinate system.

        Args:
            x (float): The X coordinate.
            y (float): The Y coordinate.
            z (float): The Z coordinate.

        Returns:
            field (np.ndarray): An array containing the magnetic field vector in
                cartesian coordinates (Bx, By, Bz)
        """
        return self._field.calc_field_xyz(x, y, z)

    def map_calc_field(self, positions):
        """
        Calculates [Br, Btheta, Bphi] for a collection of N points.

        Args:
            positions (np.ndarray): Array of spherical coordinates of shape (N, 3),
                where the last index refers to (r, theta, phi).

        Returns:
            field_rtp (np.ndarray): Array of (Br, Btheta, Bphi) of shape (N, 3)
                for each point.
        """
        return self._field.map_calc_field(
            np.asarray(positions, dtype=float).reshape(-1, 3)
        )

    def parmap_calc_field(self, positions):
        """
        Calculates [Br, Btheta, Bphi] for a collection of N points using Rayon
        for parallelization.

        Args:
            positions (np.ndarray): Array of spherical coordinates of shape (N, 3),
                where the last index refers to (r, theta, phi).

        Returns:
            field_rtp (np.ndarray): Array of (Br, Btheta, Bphi) of shape (N, 3)
                for each point.
        """
        return self._field.parmap_calc_field(
            np.asarray(positions, dtype=float).reshape(-1, 3)
        )

    def map_calc_field_xyz(self, positions):
        """
        Calculates [Bx, By, Bz] for a collection of N points. Uses cartesian coordinates
        as input.

        Args:
            positions (np.ndarray): Array of spherical coordinates of shape (N, 3),
                where the last index refers to (x, y, z) coordinates.

        Returns:
            field_xyz (np.ndarray): Array of shape (N, 3) containing (Bx, By, Bz)
                for each point.
        """
        return self._field.map_calc_field_xyz(
            np.asarray(positions, dtype=float).reshape(-1, 3)
        )

    def parmap_calc_field_xyz(self, positions):
        """
        Calculates [Bx, By, Bz] for a collection of N points. Uses cartesian coordinates
        as input and Rayon for parallelization.

        Args:
            positions (np.ndarray): Array of spherical coordinates of shape (N, 3),
                where the last index refers to (x, y, z) coordinates.

        Returns:
            field_xyz (np.ndarray): Array of shape (N, 3) containing (Bx, By, Bz)
                for each point.
        """
        return self._field.parmap_calc_field_xyz(
            np.asarray(positions, dtype=float).reshape(-1, 3)
        )

File: iupitermag-core/src/field.rs
use crate::convert;
use ndarray::{Array1, Array2, ArrayView2, Zip};

/// Contains methods that are shared between all types of magnetic field models.
pub trait Field {
    /// Calculate the field (Br, Btheta, Bphi) at a point at spherical coordinates (r, theta, phi).
    fn calc_field(&self, r: f64, theta: f64, phi: f64) -> Array1<f64>;

    /// Return the field calculated at a point (x, y, z) in cartesian coordinates
    /// (Bx, By, Bz).
    fn calc_field_xyz(&self, x: f64, y: f64, z: f64) -> Array1<f64> {
        let pos_rtp = convert::pos_xyz_to_rtp(&[x, y, z]);
        let b_rtp = self.calc_field(pos_rtp[0], pos_rtp[1], pos_rtp[2]);
        convert::vec_rtp_to_xyz(b_rtp.view(), &pos_rtp[1], &pos_rtp[2])
    }

    /// Calculate the field at a collection of points (`positions`) where `positions` is of shape
    /// (N, 3) with radius in the (N, 0),  theta in (N, 1), and phi in (N, 2).
    fn map_calc_field(&self, positions: ArrayView2<f64>) -> Array2<f64> {
        let mut result = Array2::<f64>::zeros((positions.nrows(), 3));

        Zip::from(result.rows_mut())
            .and(positions.rows())
            .for_each(|mut x, y| {
                x.assign(&self.calc_field(y[0], y[1], y[2]));
            });

        result
    }

    /// Similar to [`Field::map_calc_field`], but uses Rayon for parallelizing.
    fn parmap_calc_field(&self, positions: ArrayView2<f64>) -> Array2<f64>
    where
        Self: Sync,
    {
        let mut result = Array2::<f64>::zeros((positions.nrows(), 3));

        Zip::from(result.rows_mut())
            .and(positions.rows())
            .par_for_each(|mut x, y| {
                x.assign(&self.calc_field(y[0], y[1], y[2]));
            });

        result
    }

    /// Calculate the field at a collection of points (`positions`) where `positions` is of shape
    /// (N, 3) with the three indices indicating (X, Y, Z). Returns collection of (Bx,  By, Bz).
    fn map_calc_field_xyz(&self, positions: ArrayView2<f64>) -> Array2<f64> {
        let mut result = Array2::<f64>::zeros((positions.nrows(), 3));

        Zip::from(result.rows_mut())
            .and(positions.rows())
            .for_each(|mut x, y| {
                x.assign(&self.calc_field_xyz(y[0], y[1], y[2]));
            });

        result
    }

    /// Similar to [`Field::map_calc_field_xyz`], but uses Rayon for parallelizing.
    fn parmap_calc_field_xyz(&self, positions: ArrayView2<f64>) -> Array2<f64>
    where
        Self: Sync,
    {
        let mut result = Array2::<f64>::zeros((positions.nrows(), 3));

        Zip::from(result.rows_mut())
            .and(positions.rows())
            .par_for_each(|mut x, y| {
                x.assign(&self.calc_field_xyz(y[0], y[1], y[2]));
            });

        result
    }
}
